source/contribute/maintainer.rst

```diff
@@ -59,16 +67,167 @@ Optionally, add any additional contributor/maintainer forks, e.g.:
+
+Packages in the doctrine organisation should use following branching
+structure:
+
+ * ``develop`` - extremely unstable, points at the next planned
+   **MAJOR** release, may be rebased in order to speed up individual
+   maintainers prototyping new changes. Changes on ``develop`` can
+   be radical, and should not be relied upon.
+ * ``master`` - always to be considered as the next planned **MAJOR**
+   or **MINOR** release (depending on team internal agreement).
+   Consumers should not rely on ``master`` unless they are prepared
+   to adapt their codebase at every potentially breaking change.
+ * ``MAJOR.MINOR.x`` - always to be considered the next planned
+   **PATCH** release. Maintainers should keep these (multiple) branches
+   stable. The base of these branches MUST be the ``MAJOR.MINOR.0`` tag.
+   Without a pre-existing tag, these branches should not exist
+
+Releasing packages
+------------------
+
+**MAJOR.0.0** and **MAJOR.MINOR.0** releases **MUST** be tagged from
+``master``.
+
+When tagging a new **MAJOR.0.0** or **MAJOR.MINOR.0** release, a
+corresponding **MAJOR.MINOR.x** branch should be branched off the tag.
+
+**MAJOR.MINOR.1+** releases must be tagged from the corresponding
+``MAJOR.MINOR.x`` branch.
+
+This effectively means that a typical doctrine package should have a
+git graph like following:
+
+.. code-block:: console
+
+                                             ----- develop
+                                            /
+    1.0.0 ----- 1.1.0 ----- 2.0.0 ------ master
+      |           |           \
+      |           |            ----- 2.0.x
+      |           \
+      |            ----- 1.1.1 ----- 1.1.2 ----- 1.1.x
+      \
+       ----- 1.0.1 ----- 1.0.2 ----- 1.0.x
+
+Preparing a release
+-------------------
 
--  Topic branches **must** merge into **master** and/or any affected
-   release branches.
--  Merging a topic branch puts it into the *next* release, that is the
-   next release created from **master** and/or the next patch release
-   created from a specific release branch.
+Assuming that all tasks for a planned release are completed, a
+maintainer would then be in the position of preparing a git tag,
+which for doctrine project also corresponds to a release.
+
+To do that:
+
+-  ensure that all known introduced **BC Breaks** are documented
+   in ``UPGRADE.md``.
+-  ensure that the automated tests for the branch from which
+   a release has to be tagged are passing.
+-  prepare a release description, which should:
+    -  list all patches
+    -  describe the points of major relevance in the patch
+   maintainers may want to use a tool such
+   as `weierophinney/changelog_generator <https://github.com/weierophinney/changelog_generator>`_
+   or `jwage/changelog-generator <https://github.com/jwage/changelog-generator>`_)
+   in order to generate such release notes
+
+Then it is possible to tag a release.
+
+Please note that tags *MUST* be signed. Unsigned releases will be
```

Comment:
```suggestion
Please note that tags **MUST** be GPG-signed by either the maintainer's or Doctrine's GPG key.
Unsigned releases will be
```


As discussed, as part of automating releases. we may introduce a Doctrine GPG key that will be signed by all maintainers and internally used to create tags.

Reply:
I don't like the anonymity of a shared GPG key - anyone could sign a malicious tag and hide between the shared identity. I'd advocate against that and suggest we keep signing with individual keys.

Reply:
Wouldn't be anonymous: it would require trust from the doctrine maintainers

Reply:
Looking at one tag, would I be able to tell which maintainer signed the tag?

Reply:
No, it would be done by tooling, not by a person

Reply:
Even if it's done by tooling, as long as it's triggered by a human (`./bin/doctrine release a.b.c`) it should be signed by that human. Otherwise, there's no significant advantage to signing with the GitHub key, except that we're maintaining our own key for the sake of it.

Reply:
Key `K` is signed as trusted by maintainer `A`, `B`, `C`. Subkey `K1` is deployed to a (safe) environment, where the automation takes over tagging and releasing when things are triggered. There won't be a `./bin/doctrine release a.b.c`: the aim is to remove it, and completely automate releases as merges happen. Yes, that would be a lot of releases :)

Reply:
I don't think that's feasible **at all** at the current time, hence my opposition to this. Let's take things one at a time. If we ever get to the point where we want to do a release on every merge, we can bring this up again. Until that point, let's leave the original suggestion in: people sign tags with their GPG key instead of relying on GitHubs default key.

Reply:
> instead of relying on GitHubs default key.

Yeah, this should never be done anyway :+1: